baybe/surrogates/base.py

```diff
@@ -223,70 +253,20 @@ def fit(
                 "Continuous search spaces are currently only supported by GPs."
             )
 
-        # Create scaler objects
-        input_scaler = self._make_input_scaler(searchspace)
-        output_scaler = self._make_output_scaler(objective, measurements)
-
-        def transform_inputs(df: pd.DataFrame, /) -> pd.DataFrame:
-            """Fitted input transformation pipeline."""
-            # IMPROVE: This method currently relies on two workarounds required
-            #   due the working mechanism of sklearn's ColumnTransformer:
-            #   * Unfortunately, the transformer returns a raw array, meaning that
-            #       column names need to be manually attached afterward.
-            #   * In certain cases (e.g. in hybrid spaces), the method needs
-            #       to transform only a subset of columns. Unfortunately, it is not
-            #       possible to use a subset of columns once the transformer is set up,
-            #       which is a side-effect of the first point. As a workaround,
-            #       we thus fill the missing columns with NaN and subselect afterward.
-
-            # For the workaround, collect all comp rep columns of the parameters
-            # that are actually present in the given dataframe. At the end,
-            # we'll filter the transformed augmented dataframe down to these columns.
-            exp_rep_cols = [p.name for p in searchspace.parameters]
-            comp_rep_cols = []
-            for col in [c for c in df.columns if c in exp_rep_cols]:
-                parameter = next(p for p in searchspace.parameters if p.name == col)
-                comp_rep_cols.extend(parameter.comp_rep_columns)
-
-            # Actual workaround: augment the dataframe with NaN for missing parameters
-            df_augmented = df.reindex(columns=exp_rep_cols)
-
-            # The actual transformation step
-            out = input_scaler.transform(
-                searchspace.transform(df_augmented, allow_extra=True)
-            )
-            out = pd.DataFrame(
-                out, index=df.index, columns=input_scaler.get_feature_names_out()
-            )
-
-            # Undo the augmentation, taking into account that not all comp rep
-            # parameter columns may actually be part of the search space due
-            # to other preprocessing steps.
-            comp_rep_cols = list(set(comp_rep_cols).intersection(out.columns))
-            return out[comp_rep_cols]
-
-        def transform_outputs(df: pd.DataFrame, /) -> pd.DataFrame:
-            """Fitted output transformation pipeline."""
-            import torch
-
-            dft = objective.transform(df)
-
-            if output_scaler is _IDENTITY_TRANSFORM:
-                return dft
+        # Remember on which search space the model is trained
+        self._searchspace = searchspace
```

Comment:
Why is it necessary to explicitly remember the search space?

Reply:
Because the `posterior` method needs it for transformation. The requirements that dictate it and why design is this way:
* Our surrogate objects are searchspace-agnostic, that is, you can create a surrogate without specifying the searchspace in it's constructor (i.e. separating general object configuration from specific application context).
* Instead, the context (i.e. searchspace) is passed via the `fit` method, meaning that you can fit the same surrogate object to any searchspace. 
* But you then also need this info in the posterior computation. Since we want the `posterior` interface such that the user doesn't need to repeat providing the searchspace info (this has been fixed in the `fit` call), we don't place it into the method signature but rather access it via the stored attributed. That is, the surrogate needs to "remember" the context. 

Makes sense?

Reply:
For me, this just feels like a delayed specification of the `searchspace` - instead of "please provide already when initializing", we say "provide when calling `fit`". 
And just to be sure I understand it correctly: The reason for this design decision is just such that people do not provide the `searchspace` upon initialization but have this handled by us internally?

Reply:
Yes, there's basically two reasons why we have this design, both improving usability:
* With the given layout, you can easily create your surrogate objects irrespective of the context, i.e. you can de-/serialize them in the form of a config etc. Just imagine how a surrogate config would have to look like if you had to enter the entire searchspace config there!!
* Basically same issue but slightly different effect: You would need to specify the surrogate TWICE everywhere. That is, when you create your campaign, the campaign needs to know about the searchspace, but the campaign also needs a recommender, which in turn requires the surrogate that then expects the searchspace AGAIN. 

The current design not only solves these problems but is also much clearer to me on a semantic level: when I talk about specific specs a surrogate *has*, I would say "it is a gaussian process, with this specific kernel, etc", which is perfectly reflected by our constructor. In your way, you'd always have to add "... which is defined exactly on this particular search space, bla bla". In a paper, you'd also always hide this information in the space your "x" lives in, and define your surrogate model to be valid for all possible "x" regardless of their space. Right?

Reply:
@Scienfitz, since @AVHopp is off, I'd appreciate if you could resolve this thread if you agree.

Reply:
what we actually want to store is the transformation. The transformation cannot easily be serailzied, hence we store the searchspace from which the transformation can be inferred, so while it seems strange it seems fine to me

Reply:
Serialization is a not a problem here since its a temporary attribute that is excluded from serialization any (just like the trained model itself). So we **could** just store a lambda pointing to the transformation itself. However, I think that might make it even worse in terms of readability (plus it's all private, so we can easily change later if needed). So I'll stick with it for now